src/rsp_restspawner/spawner.py

```diff
@@ -4,40 +4,208 @@
+        (returns before creation completes)."""
+        formdata = self.options_from_form(self.user_options)
+        uname = self.user.name
+        lab_specification = {
+            "options": formdata,
+            "env": await self._get_user_environment(),
+        }
+        client = await self._configure_client()
+        r = await client.post(
+            f"{self.ctrl_url}/labs/{uname}/create",
+            json=lab_specification,
+            timeout=600.0,
+            follow_redirects=False,
+        )
+        if r.status_code == 409 or r.status_code == 303:
+            #
+            # The 409 is Conflict; so just return the Lab URL, same as
+            # we would for a new Lab.
+            #
+            # We don't actually do anything with the redirect; it returns
+            # the URL to ask the controller (as admin) for lab status.
+            #
+            # One of the objects we create (for exactly this reason) is a K8s
+            # service.  K8s will know how to route to it, so we just return
+            # its (in-cluster) DNS name as the hostname, which gets resolved
+            # to an IP address by the K8s resolver.
+            user_ns = f"{get_namespace()}-{uname}"
+            return f"http://lab.{user_ns}:8888"
+        raise SpawnerError(r)
 
     async def stop(self) -> None:
-        pass
+        client = await self._configure_client(token=self.admin_token)
+        r = await client.delete(
+            f"{self.ctrl_url}/labs/{self.user.name}", timeout=300.0
+        )
+        if r.status_code == 202 or r.status_code == 404:
+            # We're deleting it, or it wasn't there to start with.
+            return
+        raise SpawnerError(r)
 
     async def poll(self) -> Optional[int]:
         """
         Check if the pod is running.
 
         If it is, return None.  If it has exited, return the return code
         if we know it, or 1 if it exited but we don't know how.
+
+        Because we do not have direct access to the pod's exit code, we
+        are here going to return 1 for "The pod does not exist from the
+        perspective of the lab controller" and 2 for "We tried to start
+        a pod, but it failed."
+        """
+        client = await self._configure_client()
+        r = await client.get(f"{self.ctrl_url}/user-status")
+        if r.status_code == 404:
+            return 1  # No lab for user.
+        if r.status_code != 200:
+            raise SpawnerError(r)
+        result = r.json()
+        if result["status"] in (
+            LabStatus.STARTING,
+            LabStatus.RUNNING,
+            LabStatus.TERMINATING,
+        ):
+            return None
+        return 2  # Pod failed; we could check 'pod' and 'events' to see why.
+
+    async def options_form(self, spawner: Spawner) -> str:
+        if spawner != self:
+            raise RuntimeError(
+                f"options_form(): self->{self}, spawner->{spawner}"
+            )
```

Comment:
Because I couldn't think of any instance in which this method, which like all methods, has `self` as its first argument, would have any different value for the passed Spawner.  I guess maybe you could have a subclass that explicitly called super().options_from_form ?  I'll delete the check.